onnxruntime/python/tools/transformers/convert_generation.py

```diff
@@ -1272,39 +1272,96 @@ def find_past_seq_len_usage(subg: GraphProto):
     return tensor_names_to_rename, nodes_to_remove
 
 
-def replace_mha_with_gqa(model: OnnxModel, past_seq_len_input: str, kv_num_heads: int = 0, world_size: int = 1):
-    past_seq_len = past_seq_len_input
-    if past_seq_len not in model.get_graphs_input_names():
-        # Add model input for past sequence length
-        new_input = onnx.helper.make_tensor_value_info(past_seq_len, onnx.TensorProto.INT64, shape=[1])
-        model.model.graph.input.append(new_input)
+def replace_mha_with_gqa(model: OnnxModel, attn_mask: str, kv_num_heads: int = 0, world_size: int = 1):
+    # Insert attention_mask subgraph to calculate shared inputs for all GroupQueryAttention nodes
+    #
+    #                attention_mask
+    #               /              \
+    #          ReduceSum          Shape
+    #              |                |
+    #             Sub             Gather
+    #              |                |
+    #          seqlens_k   total_sequence_length
+    #              |                |
+    #        Cast to int32    Cast to int32
+
+    model.add_initializer(
+        onnx.helper.make_tensor(
```

Comment:
Why need substract one? I think ReduceSum of attention mask is the exact sequence length.

Reply:
The output of ReduceSum(attention mask) - 1 is used as indices in the GQA kernel

Reply:
The values in the output tensor from ReduceSum(attention mask) - 1 are the indices used in the GQA kernel

Reply:
seqlen_k is past sequence length. The ReduceSum gets present sequence length, -1 is to substract one new token to get past sequence length. (Note that seqlen_k is not used in prompt stage since there is no past state in prompt stage, so currently it is fine for prompt stage as well).